plugins/continuous-toolbox/src/ContinuousFlyout.ts

```diff
@@ -0,0 +1,292 @@
+        return;
+      }
+    }
+  }
+
+  /**
+   * Scrolls the flyout to given position.
+   *
+   * @param position The Y coordinate to scroll to.
+   */
+  scrollTo(position: number) {
+    // Set the scroll target to either the scaled position or the lowest
+    // possible scroll point, whichever is smaller.
+    const metrics = this.getWorkspace().getMetrics();
+    this.scrollTarget = Math.min(
+      position * this.getWorkspace().scale,
+      metrics.scrollHeight - metrics.viewHeight,
+    );
+
+    this.stepScrollAnimation();
+  }
+
+  /**
+   * Scrolls the flyout to display the given category at the top.
+   *
+   * @param category The toolbox category to scroll to in the flyout.
+   */
+  scrollToCategory(category: Blockly.ISelectableToolboxItem) {
+    const position = this.scrollPositions.get(category.getName());
+    if (!position) {
+      console.warn(`Scroll position not recorded for category ${name}`);
+      return;
+    }
+    this.scrollTo(position);
+  }
+
+  /**
+   * Step the scrolling animation by scrolling a fraction of the way to
+   * a scroll target, and request the next frame if necessary.
+   */
+  private stepScrollAnimation() {
+    if (!this.scrollTarget) return;
+
+    const currentScrollPos = -this.getWorkspace().scrollY;
+    const diff = this.scrollTarget - currentScrollPos;
+    if (Math.abs(diff) < 1) {
+      this.getWorkspace().scrollbar?.setY(this.scrollTarget);
+      this.scrollTarget = undefined;
+      return;
+    }
+    this.getWorkspace().scrollbar?.setY(
+      currentScrollPos + diff * this.scrollAnimationFraction,
+    );
+
+    requestAnimationFrame(this.stepScrollAnimation.bind(this));
+  }
+
+  /**
+   * Handles mouse wheel events.
+   *
+   * @param e The mouse wheel event to handle.
+   */
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  protected override wheel_(e: WheelEvent) {
+    // Don't scroll in response to mouse wheel events if we're currently
+    // animating scrolling to a category.
+    if (!this.scrollTarget) {
+      super.wheel_(e);
+    }
+  }
+
+  /**
+   * Calculates the additional padding needed at the bottom of the flyout in
+   * order to make it possible to scroll to the top of the last category.
+   *
+   * @param contentMetrics Content metrics for the flyout.
+   * @param viewMetrics View metrics for the flyout.
+   * @returns The additional bottom padding needed.
+   */
+  calculateBottomPadding(
+    contentMetrics: Blockly.MetricsManager.ContainerRegion,
+    viewMetrics: Blockly.MetricsManager.ContainerRegion,
+  ): number {
+    if (this.scrollPositions.size > 0) {
+      const lastPosition =
+        ([...this.scrollPositions.values()].pop() ?? 0) *
+        this.getWorkspace().scale;
+      const lastCategoryHeight = contentMetrics.height - lastPosition;
+      if (lastCategoryHeight < viewMetrics.height) {
+        return viewMetrics.height - lastCategoryHeight;
+      }
+    }
+    return 0;
```

Comment:
Guard clauses?

```suggestion
    if (this.scrollPositions.size === 0) return 0;
    const lastPosition =
      ([...this.scrollPositions.values()].pop() ?? 0) *
      this.getWorkspace().scale;
    const lastCategoryHeight = contentMetrics.height - lastPosition;
    if (lastCategoryHeight > viewMetrics.height) return 0;
    return viewMetrics.height - lastCategoryHeight;
```

Also, is the `?? 0` needed on line 218?  Surely the check on `.size` means there is guaranteed to be at least one entry?  Or is that just to satisfy the type checker?

Reply:
Done, and yes that's just there for the typechecker because pop() isn't guaranteed to return a value.